containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala

```diff
@@ -0,0 +1,346 @@
+                                   s"$sshUserName@$clipperHost") ++ startContainerCmd
+    if (sshStartContainerCmd.! != 0) {
+      throw new ModelDeploymentError("Error starting model container")
+    }
+  }
+
+  private def publishModelToClipper(host: String,
+                                    name: String,
+                                    version: Int,
+                                    labels: List[String],
+                                    hostModelDataPath: String): Unit = {
+    val data = Map(
+      "model_name" -> name,
+      "model_version" -> version,
+      "labels" -> labels,
+      "input_type" -> "doubles",
+      "container_name" -> CLIPPER_SPARK_CONTAINER_NAME,
+      "model_data_path" -> hostModelDataPath
+    )
+    val jsonData = write(data)
+
+
+    val response = Http(s"http://$host:$CLIPPER_MANAGEMENT_PORT/admin/add_model")
+      .header("Content-type", "application/json")
+      .postData(jsonData)
+      .asString
+
+    if (response.code == 200) {
+      println("Successfully published model to Clipper")
+    } else {
+      throw new ModelDeploymentError(
+        s"Error publishing model to Clipper. ${response.code}: ${response.body}")
+    }
+  }
+
+  private def copyModelDataToHost(host: String,
+                                  localPath: String,
+                                  destPath: String,
+                                  sshUserName: String,
+                                  sshKeyPath: String): Unit = {
+
+    val mkdirCommand = Seq("ssh",
+                           "-o", "StrictHostKeyChecking=no",
+                           "-i", s"$sshKeyPath",
+                           s"$sshUserName@$host",
+                           "mkdir", "-p", destPath)
+    val copyCommand = Seq("rsync",
+                          "-r",
+                          "-e", s"ssh -i $sshKeyPath -o StrictHostKeyChecking=no",
+                          s"$localPath/",
+                          s"$sshUserName@$host:$destPath")
+    if (!(mkdirCommand.! == 0 && copyCommand.! == 0)) {
+        throw new ModelDeploymentError(
+        "Error copying model data to Clipper host")
+    }
+  }
+
+  private[clipper] def saveSparkModel[M](sc: SparkContext,
+                                         name: String,
+                                         version: Int,
+                                         model: M,
+                                         containerClass: Class[_],
+                                         basePath: String): Unit = {
+    // Check that Spark is not running in the REPL
+    if (getReplOutputDir(sc).isDefined) {
+      throw new UnsupportedEnvironmentException(
+        "Clipper cannot deploy models from Spark Shell")
+    }
+    val modelPath = Paths.get(basePath, MODEL_DIRECTORY).toString
+    val modelType = model match {
+      case m: MLlibModel => {
+        m.save(sc, modelPath)
+        // Because I'm not sure how to do it in the type system, check that
```

Comment:
I actually want to leave it in there, because I'm pretty sure there is a way to do it with the type system, which would be preferable. If someone (including future me) comes along and has a bright idea about how to do that, that comment may prompt them to do something.